### ISSUE TITLE
S2583 FP: Rule is incorrectly raised on conditionally set state that carries across loop iterations (N>2)

### ISSUE THREAD

**Issue by rjgotten** [SonarSource/sonar-dotnet#8007]
### Description

S2583 is triggered erroneously on conditions inside loops that involve values across multiple iterations of the loop.

### Repro steps

```cs
private string Test(IEnumerable<int> values) 
{
  bool flag1 = false;
  bool flag2 = false;

  foreach(var value in values) 
  {
    if (value > 1) 
    {
      flag1 = true;
      continue;
    }

    if (!flag1) continue;

    flag2 = true;
  }

  if (flag2) {
    return "OK";
  }

  return "";
}
```

### Expected behavior

Rule S2583 is not triggered.

### Actual behavior

Rule S2583 is triggered on the `if (flag2)` line.

### Known workarounds

None.
Either suppress or disable the rule.

### Related information

* C#/VB.NET Plugins version: 9.8 (via SonarLint v7.3.0.77872)
* Visual Studio version: VS2022 v17.7.4
* MSBuild / dotnet version: .NET SDK 7.0.401
* SonarScanner for .NET version (if used): N/A
* Operating System: Windows 11

-------------------

Given the problems with this rule in the context of loop iterations, maybe it would be an idea to ignore variables that are defined outside the scope of a loop, but are mutated inside the loop?

Better to have a few false negatives on this rule than the continued false positives.
Handling all these false positives by suppressing them locally is going to make a mess of a project's codebase, so logically many users will simply disable the rule altogether and lose also the feedback where this rule _is_ doing the correct thing.

**Comment by gregory-paidis-sonarsource** [comment]
Hey @rjgotten ,

Thanks for raising this. 
This is a very interesting (and rare) catch you found. 

This rule, as you have astutely observed, has some issues right now, which we plan to fix in the near future, but the one you mention here is probably not going to be fixed, and it is a conscious decision.

The reason is that this is one of our most complicated (and performance-heavy) rules, and it is designed to visit a loop's body _exactly 2 times_. Your snippet is structured in such a way that the conditional-ity alters the value of `flag2` at N>2 (specifically N=3).

Unfortunately, there are 2 options:
- Visit the loop's body > 2 times and take a performance hit
- Live with some FPs like this

And we have decided the second one, because the performance hit was not acceptable.
Also, you suggested to ignore variables that are _outside-the-loop_, but there are two problems with that:
- I could rewrite your example with only loop-local variables, and still raise the FP. The only important part is how many `continue` statements you include, to short-circuit the symbolic execution >N times (N=2).
- We would lose a lot of valuable true positives. Usually in a loop, you have a mix of variables that are loop-local and "global" from the loop's perspective. E.g. the trivial sum:

```csharp
var total = 0;
foreach (var x in numbers)
{
    total += x;
}
```

I feel like most loops operate in this manner, so unfortunately the false positives of edge-case conditional `continue` is a necessary evil.

Thanks a lot for taking the time to raise this and give us your suggestions!

**Comment by gregory-paidis-sonarsource** [comment]
Hey again @rjgotten ,

I discussed it a bit internally with my team, and I will re-open this issue, in order to preserve this usecase in our backlog.
Even though it is a hard problem to solve, we acknowledge that it is painful for our users to supress these diagnostics.

Thanks a lot again :)

**Comment by Tim-Pohlmann** [comment]
This was actually solved a while ago. A similar issue can still happen for more complex cases, but this one does not raise anymore.
